ADR 011 - Optimistic Blob Size Independent Inclusion Proofs and PFB Fraud Proofs (#1040)
celestiaorg/celestia-app: docs/architecture/adr-010-optimistic-message-size-independent-inclusion-proofs-and-pfd-fraud-proofs.md

> # ADR 010: Optimistic Message Size Independent Inclusion Proofs and PFD Fraud Proofs

Reviewer: [optional]

```suggestion
# ADR 011: Optimistic Blob Size Independent Inclusion Proofs and PFB Fraud Proofs
```

- `010` => `011` to avoid a conflict with https://github.com/celestiaorg/celestia-app/pull/1005 . If you'd rather keep the 010 number, I can update https://github.com/celestiaorg/celestia-app/pull/1005 to use 011.

Accepting this suggestion implies a document wide rename of 
- `PFD` => `PFB` because https://github.com/celestiaorg/celestia-app/pull/978 merged
- `message` => `blob`

## Changelog

- 18.11.2022: Initial Draft

## Context

The verification game is between the verifier and a prover. The verifier has the message or its commitment and access to the block header of Celestia. The prover wants to prove the inclusion of the message in the DA Layer.

> Validators check if the PFD commitment matches the commitment that is referenced in [ProcessProposal](https://github.com/celestiaorg/celestia-app/blob/3473000a9ff04fccfbba83929711fe11643b782c/app/process_proposal.go#L113). If validators misbehave and coerce together to break this consensus rule, we need a fraud-proof to enforce this rule. If we assume that we have this fraud-proof then we can use the check in `ProcessProposal` to our advantage for an optimistic message size independent inclusion proof.

Reviewer: [optional] rephrase

```suggestion
Validators check if the PFB commitment matches the commitment that is referenced in [ProcessProposal](https://github.com/celestiaorg/celestia-app/blob/3473000a9ff04fccfbba83929711fe11643b782c/app/process_proposal.go#L113). Given it is possible for 2/3 validators to collude and break this consensus rule, we want a fraud-proof to inform light clients of violations to this rule. If we assume that we have this fraud-proof then we can use the check in `ProcessProposal` to our advantage for an optimistic blob size independent inclusion proof.
```

## Message size independent inclusion proof - PFD inclusion proof

Instead of proving the Merkle proof over all subtree roots over the message we create a Merkle proof over the PFD transaction as it includes the commitment already. Therefore this proof will include the shares where the PFD transaction lies in the Celestia block plus the inclusion proof of those shares to the `DataRoot`.
The verifier could do the following steps:

- Unserialize the PFD-Transaction
- verify the signatures matches the sequencer

> - verify that the nid is 0

Reviewer: ```suggestion
- verify that the namespaceID is the transaction namespaceID (i.e. `namespace.ID{0, 0, 0, 0, 0, 0, 0, 1}`)
```

ref: https://github.com/celestiaorg/celestia-core/blob/f0b9b8dfb24e2ca245d3da2f1bda460ddde1bb6c/pkg/consts/consts.go#L22
